File: app/src/debug/java/com/example/trackers/ActivityLifecycleTracker.java
package com.example.trackers;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.example.android.utils.LogUtility;

public class ActivityLifecycleTracker implements Application.ActivityLifecycleCallbacks {
	private final FragmentLifecycleTracker tracker;

	public ActivityLifecycleTracker(FragmentLifecycleTracker tracker) {
		this.tracker = tracker;
	}

	@Override
	public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
		LogUtility.d("Activity CREATED --> " + activity.getClass().getSimpleName());

		if (activity instanceof AppCompatActivity) {
			((AppCompatActivity) activity).getSupportFragmentManager()
				.registerFragmentLifecycleCallbacks(tracker, false);
		}
	}

	@Override
	public void onActivityStarted(Activity activity) {
		LogUtility.d("Activity STARTED --> " + activity.getClass().getSimpleName());
	}

	@Override
	public void onActivityResumed(Activity activity) {
		LogUtility.d("Activity RESUMED --> " + activity.getClass().getSimpleName());
	}

	@Override
	public void onActivityPaused(Activity activity) {
		LogUtility.d("Activity PAUSED --> " + activity.getClass().getSimpleName());
	}

	@Override
	public void onActivityStopped(Activity activity) {
		LogUtility.d("Activity STOPPED --> " + activity.getClass().getSimpleName());
	}

	@Override
	public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
		LogUtility.d("Activity SAVED INSTANCE STATE --> " + activity.getClass().getSimpleName());
	}

	@Override
	public void onActivityDestroyed(Activity activity) {
		LogUtility.d("Activity DESTROYED --> " + activity.getClass().getSimpleName());

		if (activity instanceof AppCompatActivity) {
			((AppCompatActivity) activity).getSupportFragmentManager()
				.unregisterFragmentLifecycleCallbacks(tracker);
		}
	}
}

File: app/src/debug/kotlin/com/example/trackers/ActivityLifecycleTracker.kt
package com.example.trackers

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.support.v7.app.AppCompatActivity

import com.example.utils.LogUtility

class ActivityLifecycleTracker(private val tracker: FragmentLifecycleTracker) : Application.ActivityLifecycleCallbacks {

  override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle) {
    LogUtility.d("Activity CREATED --> " + activity.javaClass.simpleName)

    if (activity is AppCompatActivity) {
      activity.supportFragmentManager
        .registerFragmentLifecycleCallbacks(tracker, false)
    }
  }

  override fun onActivityStarted(activity: Activity) {
    LogUtility.d("Activity STARTED --> " + activity.javaClass.simpleName)
  }

  override fun onActivityResumed(activity: Activity) {
    LogUtility.d("Activity RESUMED --> " + activity.javaClass.simpleName)
  }

  override fun onActivityPaused(activity: Activity) {
    LogUtility.d("Activity PAUSED --> " + activity.javaClass.simpleName)
  }

  override fun onActivityStopped(activity: Activity) {
    LogUtility.d("Activity STOPPED --> " + activity.javaClass.simpleName)
  }

  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
    LogUtility.d("Activity SAVED INSTANCE STATE --> " + activity.javaClass.simpleName)
  }

  override fun onActivityDestroyed(activity: Activity) {
    LogUtility.d("Activity DESTROYED --> " + activity.javaClass.simpleName)

    if (activity is AppCompatActivity) {
      activity.supportFragmentManager
        .unregisterFragmentLifecycleCallbacks(tracker)
    }
  }
}
